README.md
JIRA Worklog Sum
================

## Description

Sum your logged hours in JIRA for the current month.

## Install dependencies
```
$ pip install -r requirements.txt
```

## Usage:
```
  Options:
    -h --help
        Print help.
    -s --server http://jira.example.com REQUIRED
        JIRA server address.
    -u --username UserName REQUIRED
        Username.
    -p --password SecretPassword REQUIRED
        Password.
    -c --csv
        Print csv to stdout.
    -t --sheet
        Generate spreadsheet.
    -o --output output.xlsx
        The output file name. Must end with .xlsx!
  Usage Example:
    $ python jira_worklog_sum.py  -s https://example.jira.com -u username -p password
```

## Example output:
```
$ Total hours spent: 10.2
```
Also there should be a file in the directory you executed the script. (E.g: `jira-worklog-2018-01-01.xlsx`)

## Run tests
```
$ pytest
```

## TODO:
  * FEATURE: Query a different users worklogs with admin, with different time intervals
  * LONGTERM FEATURE: Eventually should create a REST API around it. Maybe with Flask or Django
  * Unit tests


CHANGELOG.md
# v0.4
  * Add csv feature
  * Add spreadsheet and csv flags

# v0.3
  * Add command line option flags

# v0.2
  * Add spreadsheet generation

# v0.1
  * Write total working hours to stdout


src/test_jira_worklog_sum.py
import unittest


if __name__ == '__main__':
    unittest.main(verbosity=2)


src/jiraextractor.py
from jira import JIRA
from datetime import datetime, date
from dateutil.tz import tzlocal
import dateutil.parser
import pytz


class JiraExtractor():
    def __init__(self, server, username, password, for_user, from_date, to_date):
        self.jira = JIRA(server, basic_auth=(username, password))
        self.extracted_data = self.extract_data_from_worklogs(for_user, from_date, to_date)
        self.total_time_in_seconds = self.get_worklogs_total_seconds(self.extracted_data)

    def check_dates(self, from_date, to_date):
        if from_date != 'startOfMonth()':
            vals = list(map(int, from_date.split('-')))
            try:
                date(vals[0], vals[1], vals[2])
            except IndexError as e:
                print(e)
                exit(1)

        if to_date != 'now()':
            vals = list(map(int, from_date.split('-')))
            try:
                date(vals[0], vals[1], vals[2])
            except IndexError as e:
                print(e)
                exit(1)

    def query_logged_issues(self, for_user='currentUser()', from_date='startOfMonth()', to_date='now()'):
        """ Query the issues where worklogs were made by the given user in the given
        time interval. Default is currentUser() and from the start of month until now.
        """

        self.check_dates(from_date, to_date)

        qs_user = f'worklogAuthor = {for_user} AND '
        qs_from_date = f'worklogDate >= {from_date} AND '
        qs_to_date = f'worklogDate <= {to_date} '
        qs_order = f'ORDER BY key ASC'

        logged_issues = self.jira.search_issues(qs_user + qs_from_date + qs_to_date + qs_order)
        return logged_issues

    def is_in_this_month(self, time_created):
        """ Need to check each worklog, otherwise it would sum every time spent for the
        current user. Hopefully the worklog.created is in UTC0. """

        created = dateutil.parser.parse(time_created)
        today = datetime.today()
        if created > datetime(today.year, today.month, 1, 0, 0, 0, 0, pytz.UTC):
            return True
        else:
            return False

    def get_all_worklogs_for_issues(self, issues):
        """ Extract all worklog objects into a list. Add issue key to the worklog object. """
        worklogs = []
        for issue in issues:
            for worklog in self.jira.worklogs(issue.key):
                if self.is_in_this_month(worklog.started):
                    worklog.key = issue.key
                    worklog.summary = issue.fields.summary
                    worklogs.append(worklog)
        return worklogs

    def extract_data_from_worklogs(self, for_user, from_date, to_date):
        """ Extract only the necessary attribues from the worklog objects. """
        logged_issues = self.query_logged_issues(for_user, from_date, to_date)
        worklogs = self.get_all_worklogs_for_issues(logged_issues)
        data = []
        for worklog in worklogs:
            d = {}
            d['issueKey'] = worklog.key
            d['summary'] = worklog.summary
            # The results should be in the local timezone. JIRA stores it in UTC0
            d['dayStarted'] = dateutil.parser.parse(worklog.started).astimezone(tzlocal()).day
            d['timeSpentSeconds'] = worklog.timeSpentSeconds
            data.append(d)
        return data

    def get_worklogs_total_seconds(self, data):
        """ Sum up all the time spent. """
        total = 0
        for d in data:
            total += d['timeSpentSeconds']
        return total


src/worklogmatrix.py
from itertools import groupby
from operator import itemgetter
from natsort import natsorted
from datetime import date


class WorklogMatrix:
    def __init__(self, extracted_data):
        self.number_of_rows = self.get_number_of_rows(extracted_data)
        self.number_of_columns = self.number_of_days_in_this_month()
        self.data_matrix = self.create_calendar_data_matrix(extracted_data)
        self.description_matrix = self.create_description_matrix(extracted_data)
        self.row_sums = self.get_rows_sum(self.data_matrix)
        self.col_sums = self.get_columns_sum(self.data_matrix)


    def get_uniq_keys(self, data):
        """ Create list of unique issue keys from the data. """
        uniq = []
        l = list({v['issueKey']: v for v in data}.values())
        for d in l:
            uniq.append(d['issueKey'])
        return uniq


    def get_number_of_rows(self, extracted_data):
        """ The number of tickets that have worklogs in this month by the currentUser """
        uniq_keys = self.get_uniq_keys(extracted_data)
        return len(uniq_keys)


    def sum_up_worklog_for_a_day(self, data):
        """ Key function for aggregating the time spent in an issue a day. """
        grouper = itemgetter("issueKey", "dayStarted")
        result = []
        for key, grp in groupby(sorted(data, key=grouper), grouper):
            temp_dict = dict(zip(["issueKey", "dayStarted"], key))
            temp_dict["timeSpentSeconds"] = sum(item["timeSpentSeconds"] for item in grp)
            result.append(temp_dict)
        return result


    def number_of_days_in_this_month(self):
        """ In the data matrix this equals to the number of columns """
        today = date.today()
        current_month_first_day = date(today.year, today.month, 1)
        diff = today - current_month_first_day
        return diff.days + 1


    def create_calendar_data_matrix(self, data):
        """ Create the calendar data matrix from an extracted worklog list.
            TODO: Needs to be simplified later for readability, modularity, etc.
        """

        calendar_matrix = [[0 for x in range(self.number_of_columns)] for y in range(self.number_of_rows)]
        summed_up_data = self.sum_up_worklog_for_a_day(data)
        sorted_data = natsorted(summed_up_data, key=lambda k: k['issueKey'])
        uniq_keys = self.get_uniq_keys(sorted_data)

        for i, key in enumerate(uniq_keys):
            for worklog in sorted_data:
                if worklog['issueKey'] == key:
                    calendar_matrix[int(i)][int(worklog['dayStarted'] - 1)] = round(worklog['timeSpentSeconds'] / 3600, 2)

        return calendar_matrix

    def convert_seconds_to_hours(self, data_matrix):
        """ Convert seconds to hours """
        for i, data in enumerate(data_matrix):
            for j, time in enumerate(data):
                data_matrix[i][j] = round(time / 3600, 2)
        return data_matrix

    def get_columns_sum(self, data_matrix):
        """ Sum up columns in a 2D data matrix """
        col_sums = []
        zipped = zip(*data_matrix)
        for d in zipped:
            col_sums.append(sum(d))
        return col_sums

    def get_rows_sum(self, data_matrix):
        """ Sum up rows in a 2D data matrix """
        row_sums = []
        for i in range(len(data_matrix)):
            s = sum(data_matrix[i])
            row_sums.append(s)
        return row_sums

    def create_description_matrix(self, extracted_data):
        key_and_summary = []
        for d in extracted_data:
            key_and_summary.append(tuple((d['issueKey'], d['summary'])))

        description = []
        for data in key_and_summary:
            if data in description:
                continue
            description.append(tuple((data[0], data[1])))
        return description


src/jira_worklog_sum.py
# -*- coding: utf-8 -*-
""" Sum up JIRA work logs for a user in hours
  Usage:
    python jira_worklog_sum.py https://example.jira.com username password

  Example output:
    $ Total hours spent: 10.2
"""
from jiraextractor import JiraExtractor
from worklogmatrix import WorklogMatrix

from datetime import date
import calendar
import sys
import xlsxwriter
import getopt
import csv


def usage():
    print("""Options:
    -h --help
        Print help.
    -s --server http://jira.example.com REQUIRED
        JIRA server address.
    -u --username UserName REQUIRED
        Username.
    -p --password SecretPassword REQUIRED
        Password.
    -c --csv
        Print csv to stdout.
    -t --sheet
        Generate spreadsheet.
    -o --output output.xlsx
        The output file name. Must end with .xlsx!
  Usage Example:
    $ python jira_worklog_sum.py  -s https://example.jira.com -u username -p password
  """)


def generate_spreadsheet(extracted_data, workbook_name='worklog.xlsx', start_row=0, start_col=0):
    """ Generate spreadsheet from data matrix """
    workbook = xlsxwriter.Workbook(workbook_name)
    worksheet = workbook.add_worksheet()

    worklogmatrix = WorklogMatrix(extracted_data)

    # Write headers
    headers = ['Key', 'Summary', 'Sum']
    for i, h in enumerate(headers):
        worksheet.write(start_row, start_col + i, h)
    t = date.today()
    m = calendar.month_abbr[t.month]
    for i in range(t.day): # Write month and days
        day = str(i+1).zfill(2)
        worksheet.write(start_row, 3 + i, "{0} {1}".format(m ,day))

    # Write key and summary
    desc_row_start_pos = start_row + 1
    desc_col_start_pos = start_col
    for i, desc in enumerate(worklogmatrix.description_matrix):
        for j, val in enumerate(desc):
            worksheet.write(desc_row_start_pos + i, desc_col_start_pos + j, val)
    worksheet.write(i + 2, j, 'Sum')

    # Write row totals
    row_totals_col_start_pos = desc_col_start_pos + j + 1
    row_sums = worklogmatrix.row_sums
    for i in range(worklogmatrix.number_of_rows):
        worksheet.write(desc_row_start_pos + i, row_totals_col_start_pos, row_sums[i])
    worksheet.write(desc_row_start_pos + i + 1, row_totals_col_start_pos, sum(row_sums))

    # Write data
    data_col_start_pos = row_totals_col_start_pos
    for i, data in enumerate(worklogmatrix.data_matrix):
        for j, time in enumerate(data):
            worksheet.write(desc_row_start_pos + i, data_col_start_pos + 1 + j, time)

    # Write column totals
    for i in range(worklogmatrix.number_of_columns):
        start_row_pos = desc_row_start_pos + worklogmatrix.number_of_rows
        start_col_pos = data_col_start_pos + 1 + i
        worksheet.write(start_row_pos, start_col_pos, worklogmatrix.col_sums[i])

    workbook.close()


def merge_data_and_desc(desc_matrix, data_matrix):
    merged_matrix = data_matrix

    for i, el in enumerate(desc_matrix):
        desc_list = list(el)
        merged_matrix[i].insert(0, desc_list[1])
        merged_matrix[i].insert(0, desc_list[0])

    return merged_matrix


def print_csv_data(csv_data, sep=','):
    csv_writer = csv.writer(sys.stdout, delimiter=sep, quoting=csv.QUOTE_NONNUMERIC,
        strict=True, doublequote=False, escapechar='\\')

    for i in range(len(csv_data)):
        csv_writer.writerow(csv_data[i])

def main():
    """ The main function. Parse opts/args, get jira data, generate matrix,
        csv or spreadsheet.
    """

    output_name = 'jira-worklog-' + str(date.today()) + '.xlsx'
    server_name = ''
    user_name = ''
    password = ''
    for_user = 'currentUser()'
    from_date = 'startOfMonth()'
    to_date = 'now()'
    opts = []
    output_csv = False
    output_spreadsheet = False

    try:
        opts, [] = getopt.getopt(sys.argv[1:], 'tchs:u:p:o:',
            ['sheet', 'csv', 'help', 'server=', 'username=' 'password=', 'output=',
            'foruser=', 'fromdate=', 'todate='])
    except getopt.GetoptError as err:
        print(err)
        usage()
        exit(1)

    for op, arg in opts:
        if op in ('-h', '--help'):
            usage()
            exit(1)
        elif op in ('-s', '--server'):
            server_name = arg
        elif op in ('-u', '--username'):
            user_name = arg
        elif op in ('-p', '--password'):
            password = arg
        elif op in ('-o', '--output'):
            if arg[-4:] != 'xlsx':
                print('Invalid output filetype!')
                exit(1)
            output_name = arg
        elif op in ('-c', '--csv'):
            output_csv = True
        elif op in ('-t', '--sheet'):
            output_spreadsheet = True
        elif op in ('--foruser'):
            for_user = arg
        elif op in ('--fromdate'):
            from_date = arg
        elif op in ('--todate'):
            to_date = arg
        else:
            assert False, "unhandled option"

    jira = JiraExtractor(server_name, user_name, password, for_user, from_date, to_date)
    extracted_data = jira.extracted_data
    worklog_matrix = WorklogMatrix(extracted_data)
    data_matrix = worklog_matrix.data_matrix
    description_matrix = worklog_matrix.description_matrix

    if output_spreadsheet == True:
        generate_spreadsheet(extracted_data, output_name)

    if output_csv == True:
        csv_data = merge_data_and_desc(description_matrix, data_matrix)
        print_csv_data(csv_data)

    print('Total hours spent:', jira.total_time_in_seconds / 3600)

if __name__ == '__main__':
    main()
